lib/lndk.js

```diff
@@ -0,0 +1,249 @@
+
+  const invoice_hex_str = request.startsWith('lni1') ? bech32b12.decode(request.slice(4)).toString('hex') : request
+
+  const invoice_contents = await new Promise((resolve, reject) => {
+    lndk.DecodeInvoice({
+      invoice: invoice_hex_str
+    }, (error, response) => {
+      if (error) return reject(error)
+      resolve(response)
+    })
+  })
+
+  const {
+    amount_msats,
+    description,
+    node_id,
+    chain,
+    payment_hash,
+    created_at,
+    relative_expiry,
+    features,
+    payer_note
+  } = invoice_contents
+
+  // convert from lndk response to ln-service parsePaymentRequest output layout
+  let minCltvDelta
+  for (const path of invoice_contents.payment_paths) {
+    const info = path.blinded_pay_info
+    if (minCltvDelta === undefined || info.cltv_expiry_delta < minCltvDelta) {
+      minCltvDelta = info.cltv_expiry_delta
+    }
+  }
+
+  const out = {
+    created_at: new Date(created_at * 1000).toISOString(),
+    // [chain_addresses]
+    cltv_delta: minCltvDelta,
+    description: payer_note || description,
+    // [description_hash]
+    destination: Buffer.from(node_id.key).toString('hex'),
+    expires_at: new Date((created_at + relative_expiry) * 1000).toISOString(),
+    features: features.map(bit => featureBitMap[bit]),
+    id: Buffer.from(payment_hash.hash).toString('hex'),
+    is_expired: new Date().getTime() / 1000 > created_at + relative_expiry,
+    // [metadata]
+    mtokens: '' + amount_msats,
+    network: chainsMap[chain],
+    payment: invoice_hex_str,
+    routes: invoice_contents.payment_paths.map((path) => {
+      const info = path.blinded_pay_info
+      const { introduction_node } = path.blinded_path
+      return {
+        base_fee_mtokens: '' + info.fee_base_msat,
+        cltv_delta: info.cltv_expiry_delta,
+        public_key: Buffer.from(introduction_node.node_id.key).toString('hex')
+      }
+    }),
+    safe_tokens: Math.round(toPositiveNumber(BigInt(amount_msats)) / 1000),
+    tokens: Math.floor(toPositiveNumber(BigInt(amount_msats)) / 1000),
+    bolt12: invoice_contents
+  }
+
+  return out
+}
+
+export async function fetchBolt12InvoiceFromOffer ({ lnd, offer, msats, description, timeout = 10_000 }) {
+  const lndk = lnd?.lndk
+  if (!lndk) throw new Error('lndk not installed, please use installLNDK')
+
+  return new Promise((resolve, reject) => {
+    lndk.GetInvoice({
+      offer,
+      // expects msats https://github.com/lndk-org/lndk/blob/bce93885f5fc97f3ceb15dc470117e10061de018/src/lndk_offers.rs#L182
+      amount: toPositiveNumber(msats),
+      payer_note: description,
+      response_invoice_timeout: timeout
+    }, async (error, response) => {
+      if (error) return reject(error)
+      const bech32invoice = 'lni1' + bech32b12.encode(Buffer.from(response.invoice_hex_str, 'hex'))
+
+      // sanity check
+      const parsedInvoice = await parseBolt12Request({ lnd, request: bech32invoice })
```

Comment:
i can remove the option to pass an hex string here, if you don't like supporting both representations. But since this is not just adding or removing a prefix, but doing the bech32 encoding, i'd like to keep the encoding and decoding thing here, to make it a part of the sanity check.